### PR TITLE
Hooks from version fix

### PR DESCRIPTION
**Pull request by craigh**
Define a Hook definition Spec for Core-2.0 modules. See [the spec](https://github.com/zikula/core/blob/HooksFromVersionFix/src/docs/Core-2.0/Hooks/definitions.md)

| Q                 | A
| ----------------- | ---
| Bug fix?          | yes
| New feature?      | yes
| BC breaks?        | no
| Deprecations?     | no
| Tests pass?       | yes
| Fixed tickets     | #2682, #2687 
| Refs tickets      | -
| License           | MIT
| Doc PR            | included
| Changelog updated | yes

 - implements ArrayAccess in MetaData to improve ease of access.
 - modifies how AbstractExtensionInstaller implements dependencies to use ContainerAware
 - creates a HookApi complete with docs and tests.
 - fixes BlocksModule\Helper\HookHelper to rename it to HookContainer

### CODE REVIEW

**Comment by craigh** [comment]
@cmfcmf @Guite @Kaik 

please read: https://github.com/zikula/core/blob/HooksFromVersionFix/src/docs/Core-2.0/Hooks/definitions.md

and

https://github.com/zikula/core/blob/HooksFromVersionFix/src/docs/Core-2.0/Hooks/HookContainer.md

for full explanation of this. thoughts? This solves #2682

**Comment by Guite** [comment]
Looks great :+1:

File: src/system/ExtensionsModule/Controller/AjaxController.php [hunk 1]
```diff
@@ -133,7 +133,7 @@ public function togglesubscriberareastatusAction(Request $request)
      */
     public function changeproviderareaorderAction(Request $request)
     {
-        $this->checkAjaxToken();
+//        $this->checkAjaxToken();
```
Comment (cmfcmf) [inline src/system/ExtensionsModule/Controller/AjaxController.php]: Better remove this before merging.

Comment (craigh) [inline src/system/ExtensionsModule/Controller/AjaxController.php]: yes. that was accidental. although that was a problem in my install.

**Pull request merged**